app/src/main/java/net/gsantner/markor/activity/MainActivity.java

```diff
@@ -363,17 +333,41 @@ public String getFileBrowserTitle() {
         return title;
     }
 
+    public int tabIdToPos(final int id) {
+        if(id == R.id.nav_notebook) return 0;
+        if(id == R.id.nav_todo) return 1;
+        if(id == R.id.nav_quicknote) return 2;
+        if(id == R.id.nav_more) return 3;
+        return 0;
+    }
+
+    public int getCurrentPos() {
+        return _viewPager.getCurrentItem();
+    }
+
+    public String getPosTitle(final int pos) {
+        if (pos == 0) return getFileBrowserTitle();
+        if (pos == 1) return getString(R.string.todo);
+        if (pos == 2) return getString(R.string.quicknote);
+        if (pos == 3) return getString(R.string.more);
+        return "";
+    }
+
     @OnPageChange(value = R.id.main__view_pager_container, callback = OnPageChange.Callback.PAGE_SELECTED)
     public void onViewPagerPageSelected(int pos) {
-        Menu menu = _bottomNav.getMenu();
-        PermissionChecker permc = new PermissionChecker(this);
-        (_lastBottomMenuItem != null ? _lastBottomMenuItem : menu.getItem(0)).setChecked(false);
-        _lastBottomMenuItem = menu.getItem(pos).setChecked(true);
-        updateFabVisibility(pos == 0);
-        _toolbar.setTitle(new String[]{getFileBrowserTitle(), getString(R.string.todo), getString(R.string.quicknote), getString(R.string.more)}[pos]);
+        _bottomNav.getMenu().getItem(pos).setChecked(true);
+
+        updateFabVisibility(pos == tabIdToPos(R.id.nav_notebook));
+
+        setMainTitle(getPosTitle(pos));
 
-        if (pos > 0 && pos < 3) {
-            permc.doIfExtStoragePermissionGranted(); // cannot prevent bottom tab selection
+        if (pos != tabIdToPos(R.id.nav_notebook)) {
+            restoreDefaultToolbar();
+        }
+
+        if (pos == tabIdToPos(R.id.nav_quicknote) || pos == tabIdToPos(R.id.nav_todo)) {
+            // cannot prevent bottom tab selection
+            new PermissionChecker(this).doIfExtStoragePermissionGranted();
```

Comment:
Cleanups. These seem to work fine. Will report after extensive testing.

Reply:
oh, is this maybe related/fix to my separate "observed" comment?

Reply:
To be clear, none of these changes change the functionality at all. I am just reorganizing the existing code so that things are called once, rather than 5 times. Equivalent already existed in this function in master.

Reply:
Do you like the navigation "4 bottom tabs" how it is? Or would you say, ok top tabs, or something else might fit better?

And yeah, the needed handling is really awful, there are like 5 different ways to get to one of the tabs, or 5 different callbacks when you reached one. :smiley:

Reply:
I like the tabs at the bottom. It feels clean and I can both swipe to change as well as click.

I am _really_ bad at UX design though :p. I'll need to think really hard about it and look at how others have solved similar issues before having a strong opinion.